pytensor/link/pytorch/dispatch/scalar.py

```diff
@@ -0,0 +1,328 @@
+    are_inputs_scalars = True
+    for ndim in ndims_input:
+        try:
+            if ndim > 0:
+                are_inputs_scalars = False
+        except TypeError:
+            are_inputs_scalars = False
+
+    return are_inputs_scalars
+
+
+@pytorch_funcify.register(ScalarOp)
+def pytorch_funcify_ScalarOp(op, node, **kwargs):
+    """Return pytorch function that implements the same computation as the Scalar Op.
+
+    This dispatch is expected to return a pytorch function that works on Array inputs as Elemwise does,
+    even though it's dispatched on the Scalar Op.
+    """
+
+    # We dispatch some PyTensor operators to Python operators
+    # whenever the inputs are all scalars.
+    if all_inputs_are_scalar(node):
+        pytorch_func = pytorch_funcify_scalar_op_via_py_operators(op)
+        if pytorch_func is not None:
+            return pytorch_func
+
+    nfunc_spec = getattr(op, "nfunc_spec", None)
+    if nfunc_spec is None:
+        raise NotImplementedError(f"Dispatch not implemented for Scalar Op {op}")
+
+    func_name = nfunc_spec[0]
+    print
+    # if "." in func_name:
+    #     pytorch_func = functools.reduce(getattr, [jax, *func_name.split(".")])
+    # else:
+    pytorch_func = getattr(torch, func_name)
+
+    if len(node.inputs) > op.nfunc_spec[1]:
+        # Some Scalar Ops accept multiple number of inputs, behaving as a variadic function,
+        # even though the base Op from `func_name` is specified as a binary Op.
+        # This happens with `Add`, which can work as a `Sum` for multiple scalars.
+        pytorch_variadic_func = getattr(torch, op.nfunc_variadic, None)
+        if not pytorch_variadic_func:
+            raise NotImplementedError(
+                f"Dispatch not implemented for Scalar Op {op} with {len(node.inputs)} inputs"
+            )
+
+        def pytorch_func(*args):
+            return pytorch_variadic_func(
+                torch.stack(torch.broadcast_tensors(*args), axis=0), axis=0
+            )
+
+    return pytorch_func
+
+
+@functools.singledispatch
+def pytorch_funcify_scalar_op_via_py_operators(op):
+    """Specialized JAX dispatch for Elemwise operations where all inputs are Scalar arrays.
+
+    Scalar (constant) arrays in the JAX backend get lowered to the native types (int, floats),
+    which can perform better with Python operators, and more importantly, avoid upcasting to array types
+    not supported by some JAX functions.
+    """
+    return None
+
+
+@pytorch_funcify_scalar_op_via_py_operators.register(Add)
+def pytorch_funcify_scalar_Add(op):
+    def elemwise(*inputs):
+        return sum(inputs)
+
+    return elemwise
+
+
+@pytorch_funcify_scalar_op_via_py_operators.register(Mul)
+def pytorch_funcify_scalar_Mul(op):
+    import operator
+    from functools import reduce
+
+    def elemwise(*inputs):
+        return reduce(operator.mul, inputs, 1)
+
+    return elemwise
+
+
+@pytorch_funcify_scalar_op_via_py_operators.register(Sub)
```

Comment:
most of these may not be needed in pytorch. It's a JAX specific optimization, again to try to keep python ints and floats from beginning traced